### PR TITLE
Make NoDefaultSpecified public

### LINKED ISSUE #1041: Add NoDefaultSpecified from trait_type to traits.api
This is the constant used to indicate that a trait `__init__` method is not being supplied with a default override.  This is helpful to have available for advanced users who are writing new `TraitType` subclasses and need to handle possible missing `default_value` parameters.

### PR DESCRIPTION
**Pull request by mdickinson**
We recently made `NoDefaultSpecified` private. In hindsight, that was the wrong thing to do, for various reasons:

- It's part of the signature of `TraitType` (specifically, it's the default value for the optional `default_value` argument), which is a public class that's designed to be subclassed. As such, it's a little odd to have the value private.
- The rationale for making it private was questionable: the immediate need was to fix the doc build, but the doc build only needed to be fixed by fixing the documentation itself.
- There's an open issue (#1041) to do the opposite: make `NoDefaultSpecified` truly public by adding it to `traits.api`.

This PR:
- Reverts PRs #1380 and #1378 (but keeps the extra test added in #1380)
- Renames the `NoDefaultSpecified` _class_, to avoid confusion (note that prior to #1378 and #1380, the class was immediately shadowed, so it was never publicly available)
- Adds the `NoDefaultSpecified` _instance_ to `traits.api`
- Adds tests that verify that passing a `default_value` argument of `NoDefaultSpecified` has the expected behaviour.
- Fixes the documentation to document `NoDefaultSpecified` using the `.. autodata` Sphinx directive instead of the incorrect `.. autoclass`. (This was the cause of the documentation build breakage observed in #1377.)

Closes #1041.

**Checklist**
- [x] Tests
- [x] Update API reference (`docs/source/traits_api_reference`)
- [x] Update User manual (`docs/source/traits_user_manual`)
- [x] Update type annotation hints in `traits-stubs`

### CODE REVIEW

File: traits-stubs/traits-stubs/trait_type.pyi [hunk 1]
```diff
@@ -14,6 +14,11 @@ from .base_trait_handler import BaseTraitHandler as BaseTraitHandler
 
 trait_types: Dict[str, int]
 
+
+class NoDefaultSpecified:
```
Comment (kitchoi) [inline traits-stubs/traits-stubs/trait_type.pyi:18]: `NoDefaultSpecified` is not a class. Should this be renamed too?

File: docs/source/traits_user_manual/custom.rst [hunk 2]
```diff
@@ -148,7 +158,7 @@ In order for the property to trigger notifications you must call either:
 * object.trait_property_changed(name, old, value) to not cache the value.
 * self.set_value(object, name, value) to cache the value.
 
-Likewise if the property will not be read only the get method must use 
+Likewise if the property will not be read only the get method must use
```
Comment (mdickinson) [inline docs/source/traits_user_manual/custom.rst:161]: Hmm. Drive-by cleanup of trailing whitespace performed autonomously by VS Code.

File: traits/tests/test_trait_types.py [hunk 3]
```diff
@@ -61,3 +61,35 @@ def test_numpy_validators_loaded_if_numpy_present(self):
...
+            (DefaultValue.constant, Undefined),
+        )
+
+        # Similarly, if NoDefaultSpecified is given, get_default_value()
+        # is Undefined.
+        trait_type = TraitType(default_value=NoDefaultSpecified)
+        self.assertEqual(
+            trait_type.get_default_value(),
+            (DefaultValue.constant, Undefined),
+        )
```
Comment (kitchoi) [inline traits/tests/test_trait_types.py:95]: `trait_type = TraitType(default_value=Undefined)` behaves exactly like this, too many sentinels to mean we don't have a default value it seems.

File: traits-stubs/traits-stubs/trait_type.pyi (hunk 1 above)
Comment (mdickinson) [reply traits-stubs/traits-stubs/trait_type.pyi:18]: Ah yes; that should be fixed. Not by renaming, though - the stub should be for the `NoDefaultSpecified` value, not for the class.

Comment (mdickinson) [reply traits-stubs/traits-stubs/trait_type.pyi:18]: I confess that I'm not actually sure what the right fix is, though.

File: traits/tests/test_trait_types.py (hunk 3 above)
Comment (mdickinson) [reply traits/tests/test_trait_types.py:95]: Sure. I think that's outside the scope of this PR, though. The main aim of the PR is to restore the behaviour so that there isn't a behaviour change between 6.1 and 6.2. Considering _new_ changes to behaviour can be done separately. :-)

File: traits-stubs/traits-stubs/trait_type.pyi (hunk 1 above)
Comment (mdickinson) [reply traits-stubs/traits-stubs/trait_type.pyi:18]: Given that `NoDefaultSpecified` is just a constant value, so that there are no methods, return types, etc. to record, the right fix may be simply to remove this entry from the stub. @kitchoi Opinions?

File: traits/tests/test_trait_types.py (hunk 3 above)
Comment (kitchoi) [reply traits/tests/test_trait_types.py:95]: I kind of think this is in-scope because this PR also exposes `NoDefaultSpecified` in `api` and recommends using it. But then I guess if we decided that `Undefined` and `NoDefaultSpecified` should be combined we can do `NoDefaultSpecified = Undefined` something like that later and that would be okay.

Comment (kitchoi) [reply traits/tests/test_trait_types.py:95]: Will concede that this is out of scope and the namespace can be modified later again.

File: traits-stubs/traits-stubs/trait_type.pyi (hunk 1 above)
Comment (kitchoi) [reply traits-stubs/traits-stubs/trait_type.pyi:18]: Removing sounds good to me.

Comment (mdickinson) [reply traits-stubs/traits-stubs/trait_type.pyi]: Done.

**Review by kitchoi — APPROVED** [review]
LGTM

(Since NoDefaultSpecified seems to overlap with Undefined, one of the public sentinels might simply direct to another but such can be done in a backward compatible way)
